src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/UntypedReferenceRewriter.cs

```diff
@@ -0,0 +1,108 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Health.Fhir.Core.Features.Search.Expressions;
+using Microsoft.Health.Fhir.ValueSets;
+
+namespace Microsoft.Health.Fhir.SqlServer.Features.Search.Expressions.Visitors
+{
+    /// <summary>
+    /// For expressions over a reference search parameter where the type is not specified
+    /// (e.g. "abc" instead of "Patient/123") AND the search parameter is constrained to be
+    /// of only one type, rewrites the expression to specify the single target type.
+    /// </summary>
+    internal class UntypedReferenceRewriter : ExpressionRewriterWithInitialContext<object>
+    {
+        public static readonly UntypedReferenceRewriter Instance = new UntypedReferenceRewriter();
+
+        public override Expression VisitSearchParameter(SearchParameterExpression expression, object context)
+        {
+            // Handle reference search parameters as well as composite search parameters with one or more reference components.
+            // We fist create a bitmask with bits set representing the component indexes that are candidates for this rule.
+            // Bit 0 is for reference, non-composite parameters where the reference if of one type
+            // Bits 1 and up represent the component indexes (plus one) where the component is a reference search parameter with one type
+
+            int componentCandidates = expression.Parameter.Type switch
+            {
+                SearchParamType.Reference when expression.Parameter.TargetResourceTypes?.Count == 1 => 1,
+                SearchParamType.Composite =>
+                    expression.Parameter.ResolvedComponents.Aggregate(
+                            (index: 1, flags: 0),
+                            (acc, p) => (index: acc.index + 1, flags: acc.flags | (p.Type == SearchParamType.Reference && p.TargetResourceTypes?.Count == 1 ? 1 << acc.index : 0)))
+                        .flags,
+                _ => 0
+            };
+
+            if (componentCandidates == 0)
+            {
+                // nothing to do for this expression
+                return expression;
+            }
+
+            // now see which components have a expression on the reference type
+            int componentsPresent = expression.Expression.AcceptVisitor(ParameterPredicateVisitor.Instance, null);
+
+            // now determine which ones we get a type expression
+
+            int componentsToFill = componentCandidates & ~componentsPresent;
```

Comment:
>componentCandidates & ~componentsPresent [](start = 35, length = 40)

I'm not afraid to say: "It's hard for me to understand bitwise operations"
Can we add comment about what is going on here?

Reply:
Yeah, you are right. Definitely needs a comment.

Reply:
Added an explanation. I hope it's clearer now.